### PR TITLE
Update for RTC, Timer, system ntp

### PR DESCRIPTION
**Pull request by ozarchie**

### CODE REVIEW

**Comment by ozarchie** [comment]
Update for RTC, Timer, system ntp
This update changes the way the system operates.
The Timer library is now used instead of the delay() function.
Routines are called on a timed basis as determined by the Timer library callbacks.
An RTC attached to the I2C bus provides local time.
ntp is now called once a day to sync the local RTC chip with the mqttbroker/ntp time.
(maybe this should be handled by the scheduler)

**Comment by herrold** [comment]
Routines are called on a timed basis as determined by the Timer library callbacks.

Nice !! Usually, a call-back model permits moving to a more loosely coupled, more scalable, more failure resistant and generally more resilient system

**Comment by james-prior** [comment]
The first three commits of this pull request were easy additions 
of new images and have been merged into dev.

That leaves the last commit, 3196cfd.
Wow! You did much work.
It is good to see movement toward using standard libraries instead of
[rolling our own](https://en.wikipedia.org/wiki/Not_invented_here).

I recommend changing the pull request to merge into dev instead of master.

@ozarchie wrote:

> The Timer library is now used instead of the delay() function.

Thanks! I sure don't miss delay().

I will have more to say as I study the code more.

**Comment by ozarchie** [comment]
I am not sure what you are saying so I will leave it to the github experts.
The code compiles and is currently running and addresses all the issues raised (named constants, whitespace etc) except for Lamp/Pump routineswhich I am currently updating.
There are some further changes for the sensors and the time-sync but I will not generate any further requests until you are all happy with the code, and the branch conflicts are resolved.
Cheers
Archie

**Comment by james-prior** [comment]
jp-50 resolves the merge conflicts for merging with the 
dev branch instead of merging with master branch.

I opened pull request #102 for it.
Alternatively, @ozarchie can merge my jp-50 branch on his master branch
and change the target branch from master to dev on this pull request #97.

@ozarchie can also resolve those conflicts on his own 
and ignore jp-50.

**Comment by james-prior** [comment]
This makes explicit what we should be doing
and that we have already been doing in varying degrees.

1. Make it easy for the person who handles your pull request to merge your stuff.
In particular, a pull request should not have any merge conflicts.

2. Make it easy for yourself to resolve merge conflicts.
   Let's call the branch that you want to have your branch merged into, 
   the target branch.
   For example mayaculpa's dev branch is a common target branch.
   1. Whenever the target branch changes,
   merge that target branch with your branch. The sooner one does this, 
   the smaller and easier the merge conflicts are to resolve.
   The more your branch and the target branch diverge,
   the harder it will be for you to resolve the merge conflicts.
   Make it easier on yourself by merging early and often.
   2. On one's own branch, it is better to have many small commits
   than one huge commit. When a merge with the target branch
   has many conflicts, one could abandon that particular merge,
   then attempt multiple smaller merges between the target branch
   and your older commits. 
   That can break up a big merge with many complicated conflicts
   into several smaller merges, each with less complicated conflicts.
   Also, neverminding merges, having many small commits 
   makes it easier to go back and figure out how something got screwed up
   when the inevitable bugs find you.
   3. When you want to issue a pull request,
   make sure that your branch is a child of the target branch.
   That makes the merge a trivial fast-forward merge.
   The person handling your pull request will love you for it.
   If your branch is not already a child of the target branch,
   merge the target branch into your branch.
   That makes your branch a child of the target branch.
   4. Sometimes, your pull request has no merge conflicts 
   when you create it, but has merge conflicts later. 
   This can happen when someone else's pull request is accepted before yours.
   Their changes to the target branch can create merge conflicts for your pull request.
   In that case merge the target branch with your branch again.
   Then your branch is a child of the target branch again,
   so it will not have merge conflicts and will be easy to merge.

Obviously, this is reinventing the wheel.
What web pages codify this stuff and are easy for beginners to understand?

**Comment by james-prior** [comment]
@ozarchie wrote:

> The code compiles and is currently running ...

Thanks for testing it.

Merging with master would create a new version of software 
that has not been tested, so you should change your pull request
to merge into mayaculpa's dev branch, where one expects to 
retest it.

**Comment by ozarchie** [comment]
Hi James,
The merge in dev seems to have used the polling method.
My experience is in embedded from some time ago, so I will leave it to
modern programmers to complete the coding.
Cheers Archie

**Comment by james-prior** [comment]
TL;DR: Option 2 below is the easiest way.

Thanks again to @ozarchie for switching to standard libraries for time stuff.

@ozarchie wrote:

> The merge in dev seems to have used the polling method. 

That is correct.

> My experience is in embedded from some time ago, 
> so I will leave it to modern programmers to complete the coding. 

It is already done.
ozarchie:master has already merged mayaculpa:dev.
The result is in jp-50. 
It dropped the polling of mayaculpa:dev and uses your timing stuff.
You can study the changes with the following command.

    git diff 3196cfd 630bd3f

You might need to do git fetch commands before the 630bd3f commit is available.
[meld](http://meldmerge.org/) makes the differences much easier to see.
My gmd stuff lets one use meld with git diff.
I forget which pull request or issue I posted the gmd stuff in.
@PedroSFreitas, have you had any luck using gmd?

> ... I will leave it to modern programmers to complete the coding. 

That's already done. 
There are several ways to use the completed coding,
as I wrote earlier. I elaborate below. 
The easiest is the second option.

1. Modify [this pull request #97](https://github.com/mayaculpa/hapi/pull/97).
   1. @ozarchie would accept [pull request #2](https://github.com/ozarchie/hapi/pull/2) 
   on his own github account.
   That would resolve the merge conflicts on 
   [this pull request #97](https://github.com/mayaculpa/hapi/pull/97) 
   for merging with mayaculpa:dev.
   2. @ozarchie would edit 
   [this pull request #97](https://github.com/mayaculpa/hapi/pull/97) 
   to merge into dev branch of mayaculpa (instead of master branch).
   Then the merge conflicts should disappear.
   3. The pull request handler would accept
   [this pull request #97](https://github.com/mayaculpa/hapi/pull/97).
2. The pull request handler (likely @TylerReedMC) can accept pull request #102.
Since jp-50 of pull request #102 is a child of your ozarchie:master commit,
accepting pull request #102 would also automatically resolve
[this pull request #97](https://github.com/mayaculpa/hapi/pull/97).
3. @ozarchie could do the tedious work of merging mayaculpa:dev
into ozarchive:master and ignore jp-50.
Of course @ozarchie indicated that he is not likely to do so.

The second option is the easiest.

**Comment by ozarchie** [comment]
Hi James,
I was referring to the code in my pull request.
The code in dev does not compile at all. Did you do a verify before you did
a loaded the changes? Arduino lets you verify, even without hardware.
The F() macro does not work for strcmp or strcat - just Google it.
It looks like you did a global replace of any string.
Some strings are already declared constant. Some strings are inputs to
library functions and will fail with the macro expansion.
Have you had much experience with Arduino coding?
Cheers Archie

**Comment by james-prior** [comment]
@ozarchie wrote:

> The code in dev does not compile at all. 

Thanks, that is good to know.
What is the most recent version of dev that does compile?
What is the most recent version of dev that runs?

> Did you do a verify before you did a loaded the changes? 

No. Never. Nunca.
I look forward to being able to do so
when installation instructions become available on github.
Until then, thanks for testing what I can not test.

> The F() macro does not work for strcmp or strcat - just Google it. 
> It looks like you did a global replace of any string. 

Indeed, I did. 

> Some strings are already declared constant. 

Is that incompatible with F()?
By the way, what do you mean by declared constant?
const, #define macros, or both?

> Some strings are inputs to library functions and will fail with 
> the macro expansion. Have you had much experience with Arduino coding? 

No. I have only a little experience with Arduino.

Can you enumerate when/where F() should and should not be used?
If so, please do.
We will get this stuff straightened out.
I can add changes to jp-50.

> I was referring to the code in my pull request. 

I don't understand what you are replying to.
Interleaved replying solves that ambiguity.

**Comment by james-prior** [comment]
@ozarchie wrote:

> The F() macro does not work for ...

All F() wrappers around strings were removed
in the newest commit added to jp-50.

> The code in dev does not compile at all. 

What are the first few errors in jp-50?

**Comment by james-prior** [comment]
@ozarchie wrote: 

> The code compiles and is currently running ...

Great! It is great to have a running base to work from.
What is the first child of ozarchie:master (3196cfd) toward jp-50 that does not run?

**Comment by james-prior** [comment]
Please change the merge target to mayaculpa:dev (from mayaculpa:master).

Maybe this pull request was not automatically resolved 
by accepting pull request #102 because
this pull request is trying to merge into master branch instead of dev branch.

**Comment by PedroSFreitas** [comment]
I'll try to merge manually into `dev`. If anything goes wrong, then Archie should point it to `dev`.

**Comment by james-prior** [comment]
@PedroSFreitas wrote:

> I'll try to merge manually into dev. 

You will succeed and it will be an easy fast-forward merge.
That was the point of pull request #102, 
which was been accepted/merged into (mayaculpa:)dev.

I think nothing will go wrong
and that merging 3196cfd into dev
will be insufficient to satisfy this pull request.

As long as the target of this pull request remains mayaculpa:master,
and mayaculpa:master is not a child 
of the last commit (3196cfd) of this pull request,
then I think this pull request will not be automatically satisfied.

I surmise github's behavior to be that
when the target branch of a pull request 
refers to the last commit of a pull request
or is a child of the last commit of a pull request,
then the pull request is automatically determined to be satisfied
and is closed automatically.

I see at least two ways to accomplish that for this pull request.
1. Change the target branch to (mayaculpa:)dev.
This could be done by @ozarchie, or maybe by @PedroSFreitas or @TylerReedMC.
If one can see an "edit" button near the top right of the web page
for this pull request, explore clicking the "edit" button and 
changing the target branch.
2. For the target branch to remain (mayaculpa:)master,
and for (mayaculpa:)master to be merged into (mayaculpa:)dev
or vice versa. Either way, (mayaculpa:)master would become
a child of the last commit of this pull request.

Those are my guesses. I will watch what happens.

**Comment by PedroSFreitas** [comment]
Okay, this will need some tests.

**Comment by james-prior** [comment]
Before @ozarchie added two more commits to this pull request, I wrote:

> I think nothing will go wrong

Those two extra commits have good stuff in them.
Of course, those two new commits also made this pull request not 
a parent of the then mayaculpa:dev, 
so the merge conflicts came back and things got messy.
That made the merge require work by @PedroSFreitas 
that could have been avoided.
Let's try to avoid such merge conflicts in the future.

**Comment by james-prior** [comment]
poll_on_off_thing_controller() of
src/dumb_module/arduino/hapinode/hapi_controls.ino
has an undefined variable Device
in commit 6308224
but I do not see a similar error in either parent 93b2f66 or 05e2777.
